dali/operators/audio/preemphasis_filter_op.cu

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/audio/preemphasis_filter_op.h"
+#include <vector>
+
+namespace dali {
+
+namespace detail {
+
+template <typename OutputType, typename InputType>
+struct SampleDescriptor {
+  const InputType *in;
+  OutputType *out;
+  float coeff;
+  int64_t size;
+};
+
+template <typename OutputType, typename InputType>
+void __global__ PreemphasisFilterKernel(const SampleDescriptor<OutputType, InputType> *samples) {
+  const auto &sample = samples[blockIdx.y];
+  int64_t block_size = blockDim.x;
+  int64_t block_start = block_size * blockIdx.x;
+  int64_t grid_stride = block_size * gridDim.x;
+
+  int64_t k = block_start + threadIdx.x;
+  if (k >= sample.size)
+    return;
+
+  sample.out[k] = sample.in[k] - sample.coeff * sample.in[cuda_max(0l, k-1)];
+  k += grid_stride;
+  for (; k < sample.size; k += grid_stride)
+    sample.out[k] = sample.in[k] - sample.coeff * sample.in[k-1];
+}
+
+}  // namespace detail
+
+class PreemphasisFilterGPU : public PreemphasisFilter<GPUBackend> {
+ public:
+  explicit PreemphasisFilterGPU(const OpSpec &spec) : PreemphasisFilter<GPUBackend>(spec) {}
+  void RunImpl(workspace_t<GPUBackend> &ws) override;
+
+ private:
+  template <typename OutputType, typename InputType>
+  void RunImplTyped(workspace_t<GPUBackend> &ws);
+
+  Tensor<GPUBackend> scratchpad_;
+};
+
+template <typename OutputType, typename InputType>
+void PreemphasisFilterGPU::RunImplTyped(workspace_t<GPUBackend> &ws) {
+  using SampleDesc = detail::SampleDescriptor<OutputType, InputType>;
+  const auto &input = ws.InputRef<GPUBackend>(0);
+  auto &output = ws.OutputRef<GPUBackend>(0);
+
+  std::vector<SampleDesc> samples_cpu(batch_size_);
+  for (int sample_idx = 0; sample_idx < batch_size_; sample_idx++) {
+    auto &sample = samples_cpu[sample_idx];
+    sample.in = input.tensor<InputType>(sample_idx);
+    sample.out = output.mutable_tensor<OutputType>(sample_idx);
+    sample.size = volume(input.tensor_shape(sample_idx));
+    sample.coeff = preemph_coeff_[sample_idx];
+  }
+
+  scratchpad_.set_type(TypeInfo::Create<uint8_t>());
+  int64_t sz = batch_size_ * sizeof(SampleDesc);
+  scratchpad_.Resize({sz});
```

Comment:
Can you call this resize in the constructor as well? In the case of constant batch size we would allocate memory ahead of the pipeline run, when we move to variable batch size, this reallocation would be fine.

Reply:
done